src/FSPersistor.js

```
@@ -0,0 +1,247 @@
+const fsUnlink = promisify(fs.unlink)
+const fsOpen = promisify(fs.open)
+const fsStat = promisify(fs.stat)
+const fsGlob = promisify(glob)
+
+const filterName = (key) => key.replace(/\//g, '_')
+
+async function sendFile(location, target, source) {
+  const filteredTarget = filterName(target)
+
+  // actually copy the file (instead of moving it) to maintain consistent behaviour
+  // between the different implementations
+  try {
+    const sourceStream = fs.createReadStream(source)
+    const targetStream = fs.createWriteStream(`${location}/${filteredTarget}`)
+    await pipeline(sourceStream, targetStream)
+  } catch (err) {
+    throw PersistorHelper.wrapError(
+      err,
+      'failed to copy the specified file',
+      { location, target, source },
+      WriteError
+    )
+  }
+}
+
+async function sendStream(location, target, sourceStream, sourceMd5) {
+  const fsPath = await LocalFileWriter.writeStream(sourceStream)
+  if (!sourceMd5) {
+    sourceMd5 = await _getFileMd5HashForPath(fsPath)
+  }
+
+  try {
+    await sendFile(location, target, fsPath)
+    const destMd5 = await getObjectMd5Hash(location, target)
+    if (sourceMd5 !== destMd5) {
+      await LocalFileWriter.deleteFile(`${location}/${filterName(target)}`)
+      throw new WriteError({
+        message: 'md5 hash mismatch',
+        info: { sourceMd5, destMd5, location, target }
+      })
+    }
+  } finally {
+    await LocalFileWriter.deleteFile(fsPath)
+  }
+}
+
+// opts may be {start: Number, end: Number}
+async function getObjectStream(location, name, opts) {
+  const filteredName = filterName(name)
+
+  try {
+    opts.fd = await fsOpen(`${location}/${filteredName}`, 'r')
+  } catch (err) {
+    throw PersistorHelper.wrapError(
+      err,
+      'failed to open file for streaming',
+      { location, filteredName, opts },
+      ReadError
+    )
+  }
+
+  return fs.createReadStream(null, opts)
+}
+
+async function getRedirectUrl() {
+  // not implemented
+  return null
```

maybe throw an error instead.